sql-plugin/src/main/scala/org/apache/spark/sql/rapids/RapidsShuffleInternalManagerBase.scala

```diff
@@ -63,12 +73,342 @@ abstract class GpuShuffleBlockResolverBase(
+            (0 until numPartitions).map { i =>
+              val (blockId, file) = blockManager.diskBlockManager.createTempShuffleBlock()
+              val writer: DiskBlockObjectWriter = blockManager.getDiskWriter(
+                blockId, file, serializer, fileBufferSize, writeMetrics)
+              setChecksumIfNeeded(writer, i) // spark3.2.0+
+
+              // Places writer objects at round robin slot numbers apriori
+              // this choice is for simplicity but likely needs to change so that
+              // we can handle skew better
+              val slotNum = RapidsShuffleInternalManagerBase.getNextSlot
+              diskBlockObjectWriters.put(i, (slotNum, writer))
+            }
+            writeMetrics.incWriteTime(System.nanoTime() - openStartTime);
+
+            // we call write on every writer for every record in parallel
+            val writeFutures = new ArrayBuffer[Future[Unit]]
+            records.foreach { record =>
+              val key = record._1
+              val value = record._2
+              val reducePartitionId: Int = partitioner.getPartition(key)
+              val (slotNum, myWriter) = diskBlockObjectWriters(reducePartitionId)
+
+              // we close batches actively in the `records` iterator as we get the next batch
+              // this makes sure it is kept alive while a task is able to handle it.
+              val cb = value match {
+                case columnarBatch: ColumnarBatch =>
+                  SlicedGpuColumnVector.incRefCount(columnarBatch)
+                case _ =>
+                  null
+              }
+
+              writeFutures += RapidsShuffleInternalManagerBase.queueTask(slotNum, () => {
+                withResource(cb) { _ =>
+                  myWriter.write(key, value)
+                }
+              })
+            }
+
+            withResource(new NvtxRange("WaitingForWrites", NvtxColor.PURPLE)) { _ =>
+              synchronized {
+                var errorOccurred: Throwable = null
+                while (writeFutures.nonEmpty && errorOccurred == null) {
+                  // get the head future
+                  val fut = writeFutures.remove(0)
+                  try {
+                    fut.get()
+                  } catch {
+                    case ee: ExecutionException =>
+                      // `ExecutionException` wraps the actual exception from the task,
+                      // so extract it here. This is the exception we would normally get.
+                      errorOccurred = ee.getCause
+                    case t: Throwable =>
+                      // got a different exception here, lets not ignore it
+                      errorOcurred = t
+                  }
+                }
+                if (errorOccurred != null) {
+                  // cancel all pending futures
+                  writeFutures.foreach(_.cancel(true /*ok to interrupt*/))
+                  writeFutures.clear()
```

Comment:
Nit: Not sure why this `clear` is here, since the variable is soon going out of scope due to the `throw`.

Reply:
removed